apps/designer/app/shared/design-system/components/primitives/dnd/canvas.stories.tsx

```diff
@@ -0,0 +1,308 @@
+        if (dragItemIndex !== -1) {
+          path = path.slice(dragItemIndex + 1);
+        }
+
+        for (const item of path) {
+          if (item.acceptsChildren) {
+            const element = idToElement(rootElement, item.id);
+            if (element) {
+              setDropTarget(item.id, element);
+              return;
+            }
+          }
+        }
+      }
+
+      setDropTarget(ROOT_ID, rootElement);
+    },
+  });
+
+  const autoScrollHandlers = useAutoScroll();
+
+  const dragProps = useDrag({
+    onStart(event) {
+      const id = event.target.dataset.id;
+      if (id == null) {
+        event.cancel();
+        return;
+      }
+      setDragItemId(id);
+      autoScrollHandlers.setEnabled(true);
+    },
+    onMove: (poiterCoordinate) => {
+      dropTargetHandlers.handleMove(poiterCoordinate);
+      autoScrollHandlers.handleMove(poiterCoordinate);
+      placementHandlers.handleMove(poiterCoordinate);
+    },
+    onEnd() {
+      const currentDropTargetId = dropTargetId.current;
+      if (
+        placement !== undefined &&
+        dragItemId !== undefined &&
+        currentDropTargetId !== undefined
+      ) {
+        setData((current) => {
+          const dragItem = findItem(current, dragItemId);
+
+          // shouldn't happen, for TypeScript
+          if (dragItem === undefined) {
+            return current;
+          }
+
+          return mapItems(current, (item) => {
+            let children = item.children;
+
+            const oldIndex = children.findIndex(
+              (child) => child.id === dragItemId
+            );
+
+            if (oldIndex !== -1) {
+              children = children.slice();
+              children.splice(oldIndex, 1);
+            }
+
+            if (item.id === currentDropTargetId) {
+              // placement.index does not take into account the fact that the drag item will be removed.
+              // we need to do this to account for it.
+              const newIndex =
+                oldIndex !== -1 && oldIndex < placement.index
+                  ? placement.index - 1
+                  : placement.index;
+
+              children = children.slice();
+              children.splice(newIndex, 0, dragItem);
+            }
+
+            return children === item.children ? item : { ...item, children };
+          });
+        });
+      }
+
+      dropTargetHandlers.handleEnd();
+      autoScrollHandlers.setEnabled(false);
+      placementHandlers.handleEnd();
+      setDragItemId(undefined);
+      setPalcement(undefined);
+      dropTargetId.current = undefined;
+    },
+  });
+
+  const placementHandlers = usePlacement({
+    onPalcementChange: (event) => {
+      setPalcement(event);
+    },
+  });
+
+  return (
+    <>
+      <Box
+        css={{
+          background: "white",
+          padding: 10,
+          width: 500,
+          height: 500,
+          overflow: "auto",
+          touchAction: dragItemId === undefined ? "none" : "auto",
```

Comment:
On touch devices DnD will always clash with scroll. Not sure it's possible to make it work well.

This seem to work on the demos, although I only tested in the Chrome's responsive mode.

Reply:
Sorry I messed up the condition. I was going to try `dragItemId === undefined ? "auto" : "none"`, which doesn't work.

For now the only thing that works for sure is simply `touchAction: "none"`, but it disables the scroll.

`useMove` does the same in their demo: https://react-spectrum.adobe.com/react-aria/useMove.html